d-sandbox

<div style="text-align: center; line-height: 0; padding-top: 9px;">
  <img src="https://databricks.com/wp-content/uploads/2018/03/db-academy-rgb-1200px.png" alt="Databricks Learning" style="width: 600px; height: 163px">
</div>

# Exploratory Analysis

Exploratory data analysis, or EDA, builds intuition from your data.  This lesson introduces the objectives and strategies for EDA including summary statistics, basic plotting, and correlations.

## ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) In this lesson you:<br>
* Identify the main objectives of exploratory analysis
* Calculate statistical moments to determine the center and spread of data
* Create plots of data including histograms and scatterplots
* Calculate correlations between variables
* Explore more advanced plots to visualize the relation between variables

<iframe  
src="//fast.wistia.net/embed/iframe/baptj0tmzh?videoFoam=true"
style="border:1px solid #1cb1c2;"
allowtransparency="true" scrolling="no" class="wistia_embed"
name="wistia_embed" allowfullscreen mozallowfullscreen webkitallowfullscreen
oallowfullscreen msallowfullscreen width="640" height="360" ></iframe>
<div>
<a target="_blank" href="https://fast.wistia.net/embed/iframe/baptj0tmzh?seo=false">
  <img alt="Opens in new tab" src="https://files.training.databricks.com/static/images/external-link-icon-16x16.png"/>&nbsp;Watch full-screen.</a>
</div>

-sandbox
### Exploratory Data Analysis

The goal of exploratory data analysis (EDA) is to build the intuition into a dataset that will inform how you model the data.  Models are only as strong as data that is fed into them and the more insight a data scientist has into their data, the stronger the features they'll create are and the more informed their design choices will be.

<div><img src="https://files.training.databricks.com/images/eLearning/ML-Part-1/eda.png" style="height: 500px; margin: 20px"/></div>

Exploratory analysis focuses on the following:<br><br>

1. Gain basic intuition into the data
  - What does each feature represent?
  - Are your features categorical or continuous?
  - What data needs to be encoded? (e.g. mapping a variable for gender to a number)
  - What data types are you working with? (e.g. integers, strings)
2. How is the data distributed?
  - What is the mean, median, and/or mode of each variable?
  - What is the variance, or spread, of each variable?
  - Are there outliers?  Missing values?
3. How can we visualize the data?
  - Histograms, scatterplots, and boxplots
  - Correlation plots
4. What hypotheses will I test with my models?
  - What features are correlated?
  - What are our ideal features and how can we build them if they're not already available?

Run the following cell to set up our environment.

In [6]:
%run "./Includes/Classroom-Setup"

-sandbox
### Count, Mean, and Standard Deviation

One way to start the EDA process is to calculate **n**, or the number of observations in the dataset, as well as the center and spread of the data.  We can break this down in the following way:<br><br>

1. Count gives us the number of observed values, indicating the size of the dataset and whether there are missing values.
2. Mean gives us the center of the data.  This could also be accomplished with the median or mode, depending on the data.
3. Standard deviation quantifies how spread out the data is from the mean.  A small standard deviation indicates that the data is closely centered on the mean.

<div><img src="https://files.training.databricks.com/images/eLearning/ML-Part-1/standard-deviation.png" style="height: 250px; margin: 20px"/></div>

This chart indicates how to interpret standard deviation.  68% of the data is within one standard deviation (represented by `σ`) from the mean.

<img alt="Side Note" title="Side Note" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.05em; transform:rotate(15deg)" src="https://files.training.databricks.com/static/images/icon-note.webp"/> See the <a href="https://en.wikipedia.org/wiki/Standard_deviation" target="_blank">Wikipedia article on Standard Deviation for more details on standard deviation.</a>

Import the Boston dataset and call the `.describe()` method on it.

In [9]:
bostonDF = (spark.read
  .option("HEADER", True)
  .option("inferSchema", True)
  .csv("/mnt/training/bostonhousing/bostonhousing/bostonhousing.csv")
  .drop("_c0")
)

display(bostonDF.describe())# show dataframe using .describe()

summary crim zn indus chas nox rm age dis rad tax ptratio black lstat medv count 506 506 506 506 506 506 506 506 506 506 506 506 506 506 mean 3.6135235573122535 11.363636363636363 11.136778656126504 0.0691699604743083 0.5546950592885372 6.284634387351787 68.57490118577078 3.795042687747034 9.549407114624506 408.2371541501976 18.455533596837967 356.67403162055257 12.653063241106723 22.532806324110698 stddev 8.601545105332491 23.32245299451514 6.860352940897589 0.2539940413404101 0.11587767566755584 0.7026171434153232 28.148861406903595 2.10571012662761 8.707259384239366 168.53711605495903 2.1649455237144455 91.29486438415782 7.141061511348571 9.197104087379815 min 0.00632 0.0 0.46 0 0.385 3.561 2.9 1.1296 1 187 12.6 0.32 1.73 5.0 max 88.9762 100.0 27.74 1 0.871 8.78 100.0 12.1265 24 711 22.0 396.9 37.97 50.0

This shows that there are 506 observations in our dataset with no missing values since the counts of our records are all equal.  Take a look at just the output variable `medv`.

In [11]:
# Now, display "medv"
display(bostonDF.select("medv").describe())

summary medv count 506 mean 22.532806324110698 stddev 9.197104087379815 min 5.0 max 50.0

-sandbox
As seen with the mean and standard deviation, the average house price is $22,532 with 68% of the observations within $9,197 of that average.

<img alt="Side Note" title="Side Note" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.05em; transform:rotate(15deg)" src="https://files.training.databricks.com/static/images/icon-note.webp"/> `medv` is in 1000's of dollars

-sandbox
### Plotting, Distributions, and Outliers

Anscombe's Quartet is a set of four datasets that have identical means, standard deviations, and correlations.  Despite the fact that the descriptive statistics are the same, the underlying data is very different.  This is a cautionary tale for why data should always be plotted in the EDA phase.

<div><img src="https://files.training.databricks.com/images/eLearning/ML-Part-1/anscombes_quartet.svg" style="height: 400px; margin: 20px"/></div>

Plotting of a single variable is normally first done using a histogram.  This plots the general distribution of the data with the value on the x axis and the frequency the value appears on the y axis.  Plotting of two variables is normally first done with a scatterplot.

<img alt="Side Note" title="Side Note" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.05em; transform:rotate(15deg)" src="https://files.training.databricks.com/static/images/icon-note.webp"/> For plotting a single variable in a more more rigorous way, use <a href="https://en.wikipedia.org/wiki/Kernel_density_estimation" target="_blank">kernel density estimation</a> plots instead of histograms.

-sandbox
Plot a histogram of the median value in the housing dataset.  Display this column, click on the plotting option, and visualize it as a histogram.

<div><img src="https://files.training.databricks.com/images/eLearning/ML-Part-1/plotting.png" style="height: 300px; margin: 20px"/></div>

In [15]:
display(bostonDF.select("medv"))

medv 24.0 21.6 34.7 33.4 36.2 28.7 22.9 27.1 16.5 18.9 15.0 18.9 21.7 20.4 18.2 19.9 23.1 17.5 20.2 18.2 13.6 19.6 15.2 14.5 15.6 13.9 16.6 14.8 18.4 21.0 12.7 14.5 13.2 13.1 13.5 18.9 20.0 21.0 24.7 30.8 34.9 26.6 25.3 24.7 21.2 19.3 20.0 16.6 14.4 19.4 19.7 20.5 25.0 23.4 18.9 35.4 24.7 31.6 23.3 19.6 18.7 16.0 22.2 25.0 33.0 23.5 19.4 22.0 17.4 20.9 24.2 21.7 22.8 23.4 24.1 21.4 20.0 20.8 21.2 20.3 28.0 23.9 24.8 22.9 23.9 26.6 22.5 22.2 23.6 28.7 22.6 22.0 22.9 25.0 20.6 28.4 21.4 38.7 43.8 33.2 27.5 26.5 18.6 19.3 20.1 19.5 19.5 20.4 19.8 19.4 21.7 22.8 18.8 18.7 18.5 18.3 21.2 19.2 20.4 19.3 22.0 20.3 20.5 17.3 18.8 21.4 15.7 16.2 18.0 14.3 19.2 19.6 23.0 18.4 15.6 18.1 17.4 17.1 13.3 17.8 14.0 14.4 13.4 15.6 11.8 13.8 15.6 14.6 17.8 15.4 21.5 19.6 15.3 19.4 17.0 15.6 13.1 41.3 24.3 23.3 27.0 50.0 50.0 50.0 22.7 25.0 50.0 23.8 23.8 22.3 17.4 19.1 23.1 23.6 22.6 29.4 23.2 24.6 29.9 37.2 39.8 36.2 37.9 32.5 26.4 29.6 50.0 32.0 29.8 34.9 37.0 30.5 36.4 31.1 29.1 50.0 33.3 30.3 34.6 34.9 32.9 24.1 42.3 48.5 50.0 22.6 24.4 22.5 24.4 20.0 21.7 19.3 22.4 28.1 23.7 25.0 23.3 28.7 21.5 23.0 26.7 21.7 27.5 30.1 44.8 50.0 37.6 31.6 46.7 31.5 24.3 31.7 41.7 48.3 29.0 24.0 25.1 31.5 23.7 23.3 22.0 20.1 22.2 23.7 17.6 18.5 24.3 20.5 24.5 26.2 24.4 24.8 29.6 42.8 21.9 20.9 44.0 50.0 36.0 30.1 33.8 43.1 48.8 31.0 36.5 22.8 30.7 50.0 43.5 20.7 21.1 25.2 24.4 35.2 32.4 32.0 33.2 33.1 29.1 35.1 45.4 35.4 46.0 50.0 32.2 22.0 20.1 23.2 22.3 24.8 28.5 37.3 27.9 23.9 21.7 28.6 27.1 20.3 22.5 29.0 24.8 22.0 26.4 33.1 36.1 28.4 33.4 28.2 22.8 20.3 16.1 22.1 19.4 21.6 23.8 16.2 17.8 19.8 23.1 21.0 23.8 23.1 20.4 18.5 25.0 24.6 23.0 22.2 19.3 22.6 19.8 17.1 19.4 22.2 20.7 21.1 19.5 18.5 20.6 19.0 18.7 32.7 16.5 23.9 31.2 17.5 17.2 23.1 24.5 26.6 22.9 24.1 18.6 30.1 18.2 20.6 17.8 21.7 22.7 22.6 25.0 19.9 20.8 16.8 21.9 27.5 21.9 23.1 50.0 50.0 50.0 50.0 50.0 13.8 13.8 15.0 13.9 13.3 13.1 10.2 10.4 10.9 11.3 12.3 8.8 7.2 10.5 7.4 10.2 11.5 15.1 23.2 9.7 13.8 12.7 13.1 12.5 8.5 5.0 6.3 5.6 7.2 12.1 8.3 8.5 5.0 11.9 27.9 17.2 27.5 15.0 17.2 17.9 16.3 7.0 7.2 7.5 10.4 8.8 8.4 16.7 14.2 20.8 13.4 11.7 8.3 10.2 10.9 11.0 9.5 14.5 14.1 16.1 14.3 11.7 13.4 9.6 8.7 8.4 12.8 10.5 17.1 18.4 15.4 10.8 11.8 14.9 12.6 14.1 13.0 13.4 15.2 16.1 17.8 14.9 14.1 12.7 13.5 14.9 20.0 16.4 17.7 19.5 20.2 21.4 19.9 19.0 19.1 19.1 20.1 19.9 19.6 23.2 29.8 13.8 13.3 16.7 12.0 14.6 21.4 23.0 23.7 25.0 21.8 20.6 21.2 19.1 20.6 15.2 7.0 8.1 13.6 20.1 21.8 24.5 23.1 19.7 18.3 21.2 17.5 16.8 22.4 20.6 23.9 22.0 11.9

This plot shows the general distribution of median housing prices divided into 20 bins.  Now take a look at how the number of rooms compares to the median value.  Do this by selecting the scatterplot option from the plotting dropped menu.

In [17]:
display(bostonDF.select("rm", "medv"))

rm medv 6.575 24.0 6.421 21.6 7.185 34.7 6.998 33.4 7.147 36.2 6.43 28.7 6.012 22.9 6.172 27.1 5.631 16.5 6.004 18.9 6.377 15.0 6.009 18.9 5.889 21.7 5.949 20.4 6.096 18.2 5.834 19.9 5.935 23.1 5.99 17.5 5.456 20.2 5.727 18.2 5.57 13.6 5.965 19.6 6.142 15.2 5.813 14.5 5.924 15.6 5.599 13.9 5.813 16.6 6.047 14.8 6.495 18.4 6.674 21.0 5.713 12.7 6.072 14.5 5.95 13.2 5.701 13.1 6.096 13.5 5.933 18.9 5.841 20.0 5.85 21.0 5.966 24.7 6.595 30.8 7.024 34.9 6.77 26.6 6.169 25.3 6.211 24.7 6.069 21.2 5.682 19.3 5.786 20.0 6.03 16.6 5.399 14.4 5.602 19.4 5.963 19.7 6.115 20.5 6.511 25.0 5.998 23.4 5.888 18.9 7.249 35.4 6.383 24.7 6.816 31.6 6.145 23.3 5.927 19.6 5.741 18.7 5.966 16.0 6.456 22.2 6.762 25.0 7.104 33.0 6.29 23.5 5.787 19.4 5.878 22.0 5.594 17.4 5.885 20.9 6.417 24.2 5.961 21.7 6.065 22.8 6.245 23.4 6.273 24.1 6.286 21.4 6.279 20.0 6.14 20.8 6.232 21.2 5.874 20.3 6.727 28.0 6.619 23.9 6.302 24.8 6.167 22.9 6.389 23.9 6.63 26.6 6.015 22.5 6.121 22.2 7.007 23.6 7.079 28.7 6.417 22.6 6.405 22.0 6.442 22.9 6.211 25.0 6.249 20.6 6.625 28.4 6.163 21.4 8.069 38.7 7.82 43.8 7.416 33.2 6.727 27.5 6.781 26.5 6.405 18.6 6.137 19.3 6.167 20.1 5.851 19.5 5.836 19.5 6.127 20.4 6.474 19.8 6.229 19.4 6.195 21.7 6.715 22.8 5.913 18.8 6.092 18.7 6.254 18.5 5.928 18.3 6.176 21.2 6.021 19.2 5.872 20.4 5.731 19.3 5.87 22.0 6.004 20.3 5.961 20.5 5.856 17.3 5.879 18.8 5.986 21.4 5.613 15.7 5.693 16.2 6.431 18.0 5.637 14.3 6.458 19.2 6.326 19.6 6.372 23.0 5.822 18.4 5.757 15.6 6.335 18.1 5.942 17.4 6.454 17.1 5.857 13.3 6.151 17.8 6.174 14.0 5.019 14.4 5.403 13.4 5.468 15.6 4.903 11.8 6.13 13.8 5.628 15.6 4.926 14.6 5.186 17.8 5.597 15.4 6.122 21.5 5.404 19.6 5.012 15.3 5.709 19.4 6.129 17.0 6.152 15.6 5.272 13.1 6.943 41.3 6.066 24.3 6.51 23.3 6.25 27.0 7.489 50.0 7.802 50.0 8.375 50.0 5.854 22.7 6.101 25.0 7.929 50.0 5.877 23.8 6.319 23.8 6.402 22.3 5.875 17.4 5.88 19.1 5.572 23.1 6.416 23.6 5.859 22.6 6.546 29.4 6.02 23.2 6.315 24.6 6.86 29.9 6.98 37.2 7.765 39.8 6.144 36.2 7.155 37.9 6.563 32.5 5.604 26.4 6.153 29.6 7.831 50.0 6.782 32.0 6.556 29.8 7.185 34.9 6.951 37.0 6.739 30.5 7.178 36.4 6.8 31.1 6.604 29.1 7.875 50.0 7.287 33.3 7.107 30.3 7.274 34.6 6.975 34.9 7.135 32.9 6.162 24.1 7.61 42.3 7.853 48.5 8.034 50.0 5.891 22.6 6.326 24.4 5.783 22.5 6.064 24.4 5.344 20.0 5.96 21.7 5.404 19.3 5.807 22.4 6.375 28.1 5.412 23.7 6.182 25.0 5.888 23.3 6.642 28.7 5.951 21.5 6.373 23.0 6.951 26.7 6.164 21.7 6.879 27.5 6.618 30.1 8.266 44.8 8.725 50.0 8.04 37.6 7.163 31.6 7.686 46.7 6.552 31.5 5.981 24.3 7.412 31.7 8.337 41.7 8.247 48.3 6.726 29.0 6.086 24.0 6.631 25.1 7.358 31.5 6.481 23.7 6.606 23.3 6.897 22.0 6.095 20.1 6.358 22.2 6.393 23.7 5.593 17.6 5.605 18.5 6.108 24.3 6.226 20.5 6.433 24.5 6.718 26.2 6.487 24.4 6.438 24.8 6.957 29.6 8.259 42.8 6.108 21.9 5.876 20.9 7.454 44.0 8.704 50.0 7.333 36.0 6.842 30.1 7.203 33.8 7.52 43.1 8.398 48.8 7.327 31.0 7.206 36.5 5.56 22.8 7.014 30.7 8.297 50.0 7.47 43.5 5.92 20.7 5.856 21.1 6.24 25.2 6.538 24.4 7.691 35.2 6.758 32.4 6.854 32.0 7.267 33.2 6.826 33.1 6.482 29.1 6.812 35.1 7.82 45.4 6.968 35.4 7.645 46.0 7.923 50.0 7.088 32.2 6.453 22.0 6.23 20.1 6.209 23.2 6.315 22.3 6.565 24.8 6.861 28.5 7.148 37.3 6.63 27.9 6.127 23.9 6.009 21.7 6.678 28.6 6.549 27.1 5.79 20.3 6.345 22.5 7.041 29.0 6.871 24.8 6.59 22.0 6.495 26.4 6.982 33.1 7.236 36.1 6.616 28.4 7.42 33.4 6.849 28.2 6.635 22.8 5.972 20.3 4.973 16.1 6.122 22.1 6.023 19.4 6.266 21.6 6.567 23.8 5.705 16.2 5.914 17.8 5.782 19.8 6.382 23.1 6.113 21.0 6.426 23.8 6.376 23.1 6.041 20.4 5.708 18.5 6.415 25.0 6.431 24.6 6.312 23.0 6.083 22.2 5.868 19.3 6.333 22.6 6.144 19.8 5.706 17.1 6.031 19.4 6.316 22.2 6.31 20.7 6.037 21.1 5.869 19.5 5.895 18.5 6.059 20.6 5.985 19.0 5.968 18.7 7.241 32.7 6.54 16.5 6.696 23.9 6.874 31.2 6.014 17.5 5.898 17.2 6.516 23.1 6.635 24.5 6.939 26.6 6.49 22.9 6.579 24.1 5.884 18.6 6.728 30.1 5.663 18.2 5.936 20.6 6.212 17.8 6.395 21.7 6.127 22.7 6.112 22.6 6.398 25.0 6.251 19.9 5.362 20.8 5.803 16.8 8.78 21.9 3.561 27.5 4.963

By clicking the "LOESS" box under Plot Options, a local regression line is added showing the trend.

-sandbox
Databricks notebooks can display plots generated in pure Python:<br><br>

1. Import `matplotlib`, `pandas`, and `seaborn`
2. Create a `fig` object
3. Use the `.toPandas()` DataFrame method to turn the Spark DataFrame into a Pandas DataFrame.  This way we can use Python's plotting libraries
4. Use the `scatter_matrix` pandas function to plot a matrix of scatterplots

<img alt="Caution" title="Caution" style="vertical-align: text-bottom; position: relative; height:1.3em; top:0.0em" src="https://files.training.databricks.com/static/images/icon-warning.svg"/> Do not use `.toPandas()` on large datasets since a Pandas DataFrame must fit into the driver node of a cluster.

Plot a scatter matrix using the `pandas` Python library.

In [21]:
%python
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns # importing for a better color scheme

try:
  bostonDF
except NameError: # Looks for local table if bostonDF not defined
  bostonDF = spark.table("boston")

# use matplotlib subplots
fig, ax= plt.subplots()
# create a pandas dataframe from spark dataframe, using "rm", "crim", "lstat", "medv"
pandasDF = bostonDF.select("rm","crim","lstat","medv").toPandas()
# plot the pandas dataframe using scatter matrix
pd.plotting.scatter_matrix(pandasDF)
display(fig.figure)

The above plot shows histograms on the diagonal and scatterplots in the other quadrants.  The plot in the upper-left hand quarter is a histogram of the `rm` variable and the plot to its right is a scatterplot with `rm` as the y axis and `crim` as the x axis.  This shows us correlations across our variables.

-sandbox
### Correlations

Correlation is the first way of understanding the relationship between two variables.  The correlation of two variables is the degree to which one variable increases linearly with the other:<br><br>

* A positive correlation of 1 means that for each unit increase in one variable, the same increase is seen in the other
* A negative correlation of -1 means that for each unit increase in one variable, the same decrease is seen in the other
* A correlation of 0 means that there is no association between the variables

More formally, correlation is computed as the following:<br><br>

<div><img src="https://files.training.databricks.com/images/eLearning/ML-Part-1/correlation.svg" style="height: 60px; margin: 20px"/></div>

The correlation of variables X and Y is their covariance over the standard deviation of X multiplied by the standard deviation of Y.  The covariance is calculated by summing up the product of each value minus their respective mean (`μ`).

To deepen our understanding of correlation, add columns to the `bostonDF` DataFrame that accomplishes the following:<br><br>

1. `medvX3`: `medv` multiplied by 3
3. `medvNeg`: `medv` multiplied by -3
4. `random1`: a random number
5. `random2`: a second random number
6. `medvWithNoise`: `medv` multiplied by some random noise
7. `medvWithNegativeNoise`: negative `medv` multiplied by some random noise

In [25]:
from pyspark.sql.functions import col, rand

bostonWithDummyDataDF = (bostonDF
  .select("medv")
  .withColumn("medvX3", col("medv")*3)
  .withColumn("medvNeg", col("medv")*-3)
  .withColumn("random1", rand(seed=41))
  .withColumn("random2", rand(seed=44))
  .withColumn("medvWithNoise", col("medv")*col("random1"))
  .withColumn("medvWithNegativeNoise", col("medv")*col("random1")*-1)
)

display(bostonWithDummyDataDF)

medv medvX3 medvNeg random1 random2 medvWithNoise medvWithNegativeNoise 24.0 72.0 -72.0 0.46517581696992505 0.47129200262114224 11.164219607278202 -11.164219607278202 21.6 64.80000000000001 -64.80000000000001 0.5524048845145586 0.5570435775171404 11.931945505514467 -11.931945505514467 34.7 104.10000000000001 -104.10000000000001 0.7274482336570582 0.49781132222308344 25.242453707899923 -25.242453707899923 33.4 100.19999999999999 -100.19999999999999 0.3464440249578812 0.4242408843017984 11.57123043359323 -11.57123043359323 36.2 108.60000000000001 -108.60000000000001 0.28532785423689144 0.8023509138093546 10.32886832337547 -10.32886832337547 28.7 86.1 -86.1 0.20337418139669272 0.37267849453466306 5.836839006085081 -5.836839006085081 22.9 68.69999999999999 -68.69999999999999 0.02168490133810197 0.8950894313839725 0.4965842406425351 -0.4965842406425351 27.1 81.30000000000001 -81.30000000000001 0.39986762775587437 0.742720292534395 10.836412712184195 -10.836412712184195 16.5 49.5 -49.5 0.9292327706310431 0.4281904393664595 15.332340715412212 -15.332340715412212 18.9 56.699999999999996 -56.699999999999996 0.43197171260195555 0.25760153616466275 8.16426536817696 -8.16426536817696 15.0 45.0 -45.0 0.7356681010948696 0.35851059189539713 11.035021516423043 -11.035021516423043 18.9 56.699999999999996 -56.699999999999996 0.23967816305111922 0.024723230028955934 4.529917281666153 -4.529917281666153 21.7 65.1 -65.1 0.5847410159802124 0.22536225931596965 12.68888004677061 -12.68888004677061 20.4 61.199999999999996 -61.199999999999996 0.15609694927731577 0.6627893532402052 3.1843777652572416 -3.1843777652572416 18.2 54.599999999999994 -54.599999999999994 0.5177318351636889 0.18543253196352072 9.422719399979139 -9.422719399979139 19.9 59.699999999999996 -59.699999999999996 0.16360609847615237 0.9541846784122412 3.255761359675432 -3.255761359675432 23.1 69.30000000000001 -69.30000000000001 0.5809850940379188 0.30629827555117795 13.420755672275924 -13.420755672275924 17.5 52.5 -52.5 0.6345645644201036 0.9743040803065032 11.104879877351813 -11.104879877351813 20.2 60.599999999999994 -60.599999999999994 0.3095477450659927 0.6351205875946773 6.252864450333052 -6.252864450333052 18.2 54.599999999999994 -54.599999999999994 0.3017098651787863 0.4439144042276093 5.49111954625391 -5.49111954625391 13.6 40.8 -40.8 0.9479166927396132 0.8845277125160022 12.891667021258739 -12.891667021258739 19.6 58.800000000000004 -58.800000000000004 0.4467697718338973 0.5990723063223162 8.756687527944388 -8.756687527944388 15.2 45.599999999999994 -45.599999999999994 0.8290414997123937 0.14395080644064706 12.601430795628383 -12.601430795628383 14.5 43.5 -43.5 0.40916162661782196 0.8168856973434886 5.932843585958419 -5.932843585958419 15.6 46.8 -46.8 0.37908419642977254 0.2349950192775615 5.913713464304451 -5.913713464304451 13.9 41.7 -41.7 0.933602720935892 0.4310575456903213 12.9770778210089 -12.9770778210089 16.6 49.800000000000004 -49.800000000000004 0.4208658855540822 0.13176121183114708 6.986373700197765 -6.986373700197765 14.8 44.400000000000006 -44.400000000000006 0.3553321412481544 0.576291632666724 5.258915690472685 -5.258915690472685 18.4 55.199999999999996 -55.199999999999996 0.227088740467838 0.8305360186871418 4.178432824608219 -4.178432824608219 21.0 63.0 -63.0 0.14768396476956325 0.8020806576050306 3.101363260160828 -3.101363260160828 12.7 38.099999999999994 -38.099999999999994 0.7403276197389179 0.5763096922160267 9.402160770684256 -9.402160770684256 14.5 43.5 -43.5 0.4309088620259196 0.05749284984541292 6.248178499375834 -6.248178499375834 13.2 39.599999999999994 -39.599999999999994 0.8127662718332773 0.43608116853153156 10.72851478819926 -10.72851478819926 13.1 39.3 -39.3 0.686189877996228 0.013001278152443896 8.989087401750586 -8.989087401750586 13.5 40.5 -40.5 0.9081113307525643 0.6306409637042883 12.25950296515962 -12.25950296515962 18.9 56.699999999999996 -56.699999999999996 0.9024848196206428 0.8694763270519723 17.05696309083015 -17.05696309083015 20.

Calculate correlations between all columns and `id`

In [27]:
for col in bostonWithDummyDataDF.columns:
  #TODO use corr function of stat to find out the correlation
  correlation = bostonWithDummyDataDF.stat.corr("medv",col)
  
  print("The correlation between columns 'id' and '{}': \t{}".format(col, correlation))

The correlation between columns 'id' and 'medv': 1.0
The correlation between columns 'id' and 'medvX3': 1.0000000000000004
The correlation between columns 'id' and 'medvNeg': -1.0000000000000004
The correlation between columns 'id' and 'random1': -0.05136186927895015
The correlation between columns 'id' and 'random2': -0.01392812761724272
The correlation between columns 'id' and 'medvWithNoise': 0.5089996917519524
The correlation between columns 'id' and 'medvWithNegativeNoise': -0.5089996917519524

The main takeaways:<br><br>

* The correlation between `id` and itself is 1, as is the correlation between `id` and `id` multiplied by 3.  
* The correlation between `id` and `id` multiplied by -3 is -1.  
* The correlations between `id` and random noise varied based on what was in the noise, but the results were closer to 0.  
* There were stronger positive and negative correlations in random noise multiplied by `id` than in random noise not correlated to `id`

### Other Visualization Tools

There are a number of other helpful visualizations depending on the needs of your data.  These include:<br><br>

* <a href="https://en.wikipedia.org/wiki/Heat_map" target="_blank">Heat maps:</a> similar to a scattermatrix, heatmaps can be especially helpful at visualizing correlations between variables
* <a href="https://en.wikipedia.org/wiki/Box_plot" target="_blank">Box plots:</a> visualizes quantiles and outliers
* <a href="https://en.wikipedia.org/wiki/Q%E2%80%93Q_plot" target="_blank">Q-Q Plots:</a> visualizes two probability distributions
* <a href="https://en.wikipedia.org/wiki/Geographic_information_system" target="_blank">Maps and GIS:</a> visualizes geographically-bound data
* <a href="https://en.wikipedia.org/wiki/T-distributed_stochastic_neighbor_embedding" target="_blank">t-SNE:</a> plots high dimensional data (i.e. data that has many variables) by projecting it down into two-diminsional plot
* <a href="https://en.wikipedia.org/wiki/Time_series" target="_blank">Time series:</a> plots time-bound variables including run charts, lag plots, and wavelet spectrograms

Assemble all of the `bostonDF` features into a single column `features`.  This allows us to use Spark's built-in correlation functionality.

In [31]:
from pyspark.ml.feature import VectorAssembler

# Get ALL columns in the boston dataframe to VectorAssembler
assembler=VectorAssembler(inputCols=bostonDF.columns,outputCol="Features")
# create the input of the model
bostonFeaturizedDF = assembler.transform(bostonDF)

Calculate the correlations across the entire dataset.

In [33]:
from pyspark.ml.stat import Correlation

pearsonCorr = Correlation.corr(bostonFeaturizedDF,"features").collect()[0][0]
pandasDF = pd.DataFrame(pearsonCorr.toArray())#TODO

pandasDF.index, pandasDF.columns = bostonDF.columns, bostonDF.columns # Labels our index and columns so we can interpret the results

Plot a heatmap of the correlations.  The redder the value, the stronger the positive correlation and the bluer the value the stronger the negative correlation.  Do this in pure Python rather than Spark because Python has better plotting functionality.

In [35]:
import matplotlib.pyplot as plt
import seaborn as sns

fig, ax = plt.subplots()
# plot pandas dataframe using heatmap in seaborn
sns.heatmap(pandasDF)
display(fig.figure)

## Exercise: EDA on the Bike Sharing Dataset

Do exploratory analysis on the bike sharing dataset by calculating and interpreting summary statistics, creating basic plots, and calculating correlations.

### Step 1: Summary Statistics

Calculate the count, mean, and standard deviation for each variable in the dataset.  What does each variable signify?  What is the spread of the data?

Import the data.

In [39]:
bikeDF = (spark
  .read
  .option("header", True)
  .option("inferSchema", True)
  .csv("/mnt/training/bikeSharing/data-001/hour.csv")
  .drop("instant", "dteday", "casual", "registered", "holiday", "weekday")
)
display(bikeDF)

season yr mnth hr workingday weathersit temp atemp hum windspeed cnt 1 0 1 0 0 1 0.24 0.2879 0.81 0.0 16 1 0 1 1 0 1 0.22 0.2727 0.8 0.0 40 1 0 1 2 0 1 0.22 0.2727 0.8 0.0 32 1 0 1 3 0 1 0.24 0.2879 0.75 0.0 13 1 0 1 4 0 1 0.24 0.2879 0.75 0.0 1 1 0 1 5 0 2 0.24 0.2576 0.75 0.0896 1 1 0 1 6 0 1 0.22 0.2727 0.8 0.0 2 1 0 1 7 0 1 0.2 0.2576 0.86 0.0 3 1 0 1 8 0 1 0.24 0.2879 0.75 0.0 8 1 0 1 9 0 1 0.32 0.3485 0.76 0.0 14 1 0 1 10 0 1 0.38 0.3939 0.76 0.2537 36 1 0 1 11 0 1 0.36 0.3333 0.81 0.2836 56 1 0 1 12 0 1 0.42 0.4242 0.77 0.2836 84 1 0 1 13 0 2 0.46 0.4545 0.72 0.2985 94 1 0 1 14 0 2 0.46 0.4545 0.72 0.2836 106 1 0 1 15 0 2 0.44 0.4394 0.77 0.2985 110 1 0 1 16 0 2 0.42 0.4242 0.82 0.2985 93 1 0 1 17 0 2 0.44 0.4394 0.82 0.2836 67 1 0 1 18 0 3 0.42 0.4242 0.88 0.2537 35 1 0 1 19 0 3 0.42 0.4242 0.88 0.2537 37 1 0 1 20 0 2 0.4 0.4091 0.87 0.2537 36 1 0 1 21 0 2 0.4 0.4091 0.87 0.194 34 1 0 1 22 0 2 0.4 0.4091 0.94 0.2239 28 1 0 1 23 0 2 0.46 0.4545 0.88 0.2985 39 1 0 1 0 0 2 0.46 0.4545 0.88 0.2985 17 1 0 1 1 0 2 0.44 0.4394 0.94 0.2537 17 1 0 1 2 0 2 0.42 0.4242 1.0 0.2836 9 1 0 1 3 0 2 0.46 0.4545 0.94 0.194 6 1 0 1 4 0 2 0.46 0.4545 0.94 0.194 3 1 0 1 6 0 3 0.42 0.4242 0.77 0.2985 2 1 0 1 7 0 2 0.4 0.4091 0.76 0.194 1 1 0 1 8 0 3 0.4 0.4091 0.71 0.2239 8 1 0 1 9 0 2 0.38 0.3939 0.76 0.2239 20 1 0 1 10 0 2 0.36 0.3485 0.81 0.2239 53 1 0 1 11 0 2 0.36 0.3333 0.71 0.2537 70 1 0 1 12 0 2 0.36 0.3333 0.66 0.2985 93 1 0 1 13 0 2 0.36 0.3485 0.66 0.1343 75 1 0 1 14 0 3 0.36 0.3485 0.76 0.194 59 1 0 1 15 0 3 0.34 0.3333 0.81 0.1642 74 1 0 1 16 0 3 0.34 0.3333 0.71 0.1642 76 1 0 1 17 0 1 0.34 0.3333 0.57 0.194 65 1 0 1 18 0 2 0.36 0.3333 0.46 0.3284 53 1 0 1 19 0 1 0.32 0.2879 0.42 0.4478 30 1 0 1 20 0 1 0.3 0.2727 0.39 0.3582 22 1 0 1 21 0 1 0.26 0.2273 0.44 0.3284 31 1 0 1 22 0 1 0.24 0.2121 0.44 0.2985 9 1 0 1 23 0 1 0.22 0.2273 0.47 0.1642 8 1 0 1 0 1 1 0.22 0.197 0.44 0.3582 5 1 0 1 1 1 1 0.2 0.1667 0.44 0.4179 2 1 0 1 4 1 1 0.16 0.1364 0.47 0.3881 1 1 0 1 5 1 1 0.16 0.1364 0.47 0.2836 3 1 0 1 6 1 1 0.14 0.1061 0.5 0.3881 30 1 0 1 7 1 1 0.14 0.1364 0.5 0.194 64 1 0 1 8 1 1 0.14 0.1212 0.5 0.2836 154 1 0 1 9 1 1 0.16 0.1364 0.43 0.3881 88 1 0 1 10 1 1 0.18 0.1667 0.43 0.2537 44 1 0 1 11 1 1 0.2 0.1818 0.4 0.3284 51 1 0 1 12 1 1 0.22 0.2121 0.35 0.2985 61 1 0 1 13 1 1 0.24 0.2121 0.35 0.2836 61 1 0 1 14 1 1 0.26 0.2424 0.3 0.2836 77 1 0 1 15 1 1 0.26 0.2424 0.3 0.2537 72 1 0 1 16 1 1 0.26 0.2424 0.3 0.2537 76 1 0 1 17 1 1 0.24 0.2273 0.3 0.2239 157 1 0 1 18 1 1 0.24 0.2576 0.32 0.1045 157 1 0 1 19 1 1 0.2 0.2576 0.47 0.0 110 1 0 1 20 1 1 0.2 0.2273 0.47 0.1045 52 1 0 1 21 1 1 0.18 0.197 0.64 0.1343 52 1 0 1 22 1 1 0.14 0.1515 0.69 0.1343 20 1 0 1 23 1 1 0.18 0.2121 0.55 0.1045 12 1 0 1 0 1 1 0.16 0.1818 0.55 0.1045 5 1 0 1 1 1 1 0.16 0.1818 0.59 0.1045 2 1 0 1 2 1 1 0.14 0.1515 0.63 0.1343 1 1 0 1 4 1 1 0.14 0.1818 0.63 0.0896 2 1 0 1 5 1 1 0.12 0.1515 0.68 0.1045 4 1 0 1 6 1 1 0.12 0.1515 0.74 0.1045 36 1 0 1 7 1 1 0.12 0.1515 0.74 0.1343 94 1 0 1 8 1 1 0.14 0.1515 0.69 0.1642 179 1 0 1 9 1 1 0.16 0.1515 0.64 0.2239 100 1 0 1 10 1 2 0.16 0.1364 0.69 0.3284 42 1 0 1 11 1 1 0.22 0.2121 0.51 0.2985 57 1 0 1 12 1 1 0.22 0.2273 0.51 0.1642 78 1 0 1 13 1 1 0.24 0.2273 0.56 0.194 97 1 0 1 14 1 1 0.26 0.2576 0.52 0.2239 63 1 0 1 15 1 1 0.28 0.2727 0.52 0.2537 65 1 0 1 16 1 1 0.3 0.2879 0.49 0.2537 83 1 0 1 17 1 1 0.28 0.2727 0.48 0.2239 212 1 0 1 18 1 1 0.26 0.2576 0.48 0.194 182 1 0 1 19 1 1 0.24 0.2576 0.48 0.1045 112 1 0 1 20 1 1 0.24 0.2576 0.48 0.1045 54 1 0 1 21 1 1 0.22 0.2727 0.64 0.0 48 1 0 1 22 1 1 0.22 0.2576 0.64 0.0896 35 1 0 1 23 1 1 0.2 0.2273 0.69 0.0896 11 1 0 1 0 1 1 0.2 0.2576 0.64 0.0 6 1 0 1 1 1 1 0.16 0.197 0.74 0.0896 6 1 0 1 2 1 1 0.16 0.197 0.74 0.0896 2 1 0 1 4 1 1 0.24 0.2273 0.48 0.2239 2 1 0 1 5 1 1 0.22 0.2273 0.47 0.1642 3 1 0 1 6 1 1 0.2 0.197 0.47 0.2239 33 1 0 1 7 1 1 0.18 0.1818 0.43 0.194 88 1 0 1 8 1 1 0.2 0.1818 0.4 0.2985 195 1 0 1 9 1 1 0.22 0.197 0.37 0.3284 115 1 0 1 10 1 1 0.22 0.197 0.37 0.

Calculate count, mean, and standard deviation.

In [41]:
# TODO
display(bikeDF.select("*").describe())

summary season yr mnth hr workingday weathersit temp atemp hum windspeed cnt count 17379 17379 17379 17379 17379 17379 17379 17379 17379 17379 17379 mean 2.5016399102364923 0.5025605615973301 6.537775476149376 11.546751826917545 0.6827205247712756 1.425283387997008 0.4969871684216586 0.4757751021347581 0.6272288394038822 0.1900976063064631 189.46308763450142 stddev 1.10691813944808 0.5000078290910193 3.4387757137501724 6.9144050952644776 0.4654306335238818 0.6393568777542525 0.19255612124972202 0.1718502156353594 0.19292983406291458 0.12234022857279034 181.3875990918646 min 1 0 1 0 0 1 0.02 0.0 0.0 0.0 1 max 4 1 12 23 1 4 1.0 1.0 1.0 0.8507 977

### Step 2: Plotting

Create the following plots:<br><br>

1. A histogram of the dependent variable `cnt`
2. A barplot of counts by hour
3. A scattermatrix

Create a histogram of the variable `cnt`.

In [44]:
# TODO
display(bikeDF.select("cnt"))

cnt 16 40 32 13 1 1 2 3 8 14 36 56 84 94 106 110 93 67 35 37 36 34 28 39 17 17 9 6 3 2 1 8 20 53 70 93 75 59 74 76 65 53 30 22 31 9 8 5 2 1 3 30 64 154 88 44 51 61 61 77 72 76 157 157 110 52 52 20 12 5 2 1 2 4 36 94 179 100 42 57 78 97 63 65 83 212 182 112 54 48 35 11 6 6 2 2 3 33 88 195 115 57 46 79 71 62 62 89 190 169 132 89 43 42 19 11 4 2 1 4 36 95 219 122 45 59 84 67 70 62 86 172 163 112 69 48 52 23 17 7 1 1 5 34 84 210 134 63 67 59 73 50 72 87 187 123 95 51 39 36 15 25 16 16 7 1 5 2 9 15 20 61 62 98 102 95 74 76 69 55 30 28 37 34 22 25 12 11 4 1 1 1 6 10 19 49 49 83 75 72 82 92 62 48 41 38 20 15 6 5 1 3 1 3 3 31 77 188 94 31 30 52 54 47 45 74 178 155 95 74 38 24 18 12 3 3 6 27 99 217 130 54 35 57 52 63 47 76 136 95 51 32 20 29 19 7 6 1 5 16 54 128 81 39 35 55 49 44 49 68 139 137 83 56 57 33 20 7 2 2 3 4 3 28 72 202 139 38 37 52 83 42 60 78 162 144 99 64 40 30 15 14 5 1 1 8 17 70 158 117 44 53 61 77 64 68 90 159 139 92 68 52 36 27 28 20 12 8 5 1 3 10 23 33 59 72 89 101 118 129 128 83 84 74 41 57 26 44 39 23 16 15 1 2 1 3 18 32 79 93 104 118 91 113 99 105 67 61 57 28 21 18 17 16 8 2 3 1 5 13 33 47 57 64 80 93 86 93 82 71 92 60 33 27 13 4 3 22 28 35 61 125 133 99 83 41 33 20 3 7 3 2 7 32 90 197 109 47 52 70 78 75 82 104 197 161 112 76 59 59 28 13 5 2 1 1 6 35 101 249 143 57 68 84 98 81 70 91 215 185 152 126 57 56 31 21 6 2 1 1 5 27 68 217 166 63 59 78 73 62 65 97 161 120 96 53 41 34 27 13 12 11 7 3 2 8 27 40 53 63 70 84 75 103 83 67 54 59 45 39 30 33 22 13 18 5 3 1 2 19 28 58 99 116 87 110 77 65 55 49 50 35 25 28 21 7 1 1 1 5 15 84 177 102 40 46 63 60 45 57 70 184 153 106 81 59 35 24 9 5 2 1 9 36 108 238 144 55 61 106 93 68 84 116 222 225 146 119 45 53 40 17 5 10 1 8 30 72 58 28 41 48 47 36 43 36 26 24 84 104 79 59 38 27 16 9 3 2 1 4 16 60 157 101 49 30 29 31 38 41 80 149 109 89 62 58 26 23 28 20 15 8 3 2 5 34 34 55 64 78 65 99 120 107 91 68 58 43 36 32 33 33 29 11 8 1 3 3 12 38 64 59 97 84 122 109 123 77 65 55 33 28 21 21 7 7 1 2 2 8 37 72 185 112 69 48 68 54 86 44 86 161 156 111 78 56 34 17 8 3 2 2 3 22 52 135 116 47 51 55 52 54 52 64 176 168 108 74 64 36 16 2 3 4 1 1 3 18 49 155 123 61 52 64 75 63 76 103 190 182 91 75 63 40 32 12 5 2 1 2 39 87 188 133 52 64 69 51 47 60 78 175 147 96 109 54 41 38 13 7 1 1 7 28 87 220 127 51 64 86 82 91 90 99 205 155 103 71 43 46 31 39 18 17 11 8 9 4 4 10 20 34 47 52 72 55 60 71 78 83 84 69 56 45 59 39 44 20 13 2 1 1 8 23 45 89 117 174 182 161 182 157 121 78 21 26 27 62 30 15 5 3 1 2 10 30 95 230 118 55 47 66 64 60 50 114 216 175 128 88 78 37 25 10 4 2 1 3 2 39 100 243 135 48 50 65 50 64 51 83 176 152 101 56 54 29 12 17 7 3 2 7 43 99 199 123 59 41 61 69 58 64 79 166 170 88 84 83 46 37 16 7 3 1 6 26 99 178 122 35 45 69 62 48 50 80 165 160 112 97 72 51 34 14 7 3 4 24 74 216 140 44 64 71 110 84 74 125 211 174 101 63 47 43 53 30 24 15 10 4 1 2 11 30 43 84 114 120 135 120 174 145 137 64 41 40 51 45 32 39 24 20 12 2 5 3 12 47 105 112 152 154 161 162 134 125 95 61 47 51 36 30 11 7 2 2 2 3 26 98

Create a barplot of counts by hour.

In [46]:
display(bikeDF.select("hr","cnt"))

hr cnt 0 16 1 40 2 32 3 13 4 1 5 1 6 2 7 3 8 8 9 14 10 36 11 56 12 84 13 94 14 106 15 110 16 93 17 67 18 35 19 37 20 36 21 34 22 28 23 39 0 17 1 17 2 9 3 6 4 3 6 2 7 1 8 8 9 20 10 53 11 70 12 93 13 75 14 59 15 74 16 76 17 65 18 53 19 30 20 22 21 31 22 9 23 8 0 5 1 2 4 1 5 3 6 30 7 64 8 154 9 88 10 44 11 51 12 61 13 61 14 77 15 72 16 76 17 157 18 157 19 110 20 52 21 52 22 20 23 12 0 5 1 2 2 1 4 2 5 4 6 36 7 94 8 179 9 100 10 42 11 57 12 78 13 97 14 63 15 65 16 83 17 212 18 182 19 112 20 54 21 48 22 35 23 11 0 6 1 6 2 2 4 2 5 3 6 33 7 88 8 195 9 115 10 57 11 46 12 79 13 71 14 62 15 62 16 89 17 190 18 169 19 132 20 89 21 43 22 42 23 19 0 11 1 4 2 2 4 1 5 4 6 36 7 95 8 219 9 122 10 45 11 59 12 84 13 67 14 70 15 62 16 86 17 172 18 163 19 112 20 69 21 48 22 52 23 23 0 17 1 7 2 1 4 1 5 5 6 34 7 84 8 210 9 134 10 63 11 67 12 59 13 73 14 50 15 72 16 87 17 187 18 123 19 95 20 51 21 39 22 36 23 15 0 25 1 16 2 16 3 7 4 1 5 5 6 2 7 9 8 15 9 20 10 61 11 62 12 98 13 102 14 95 15 74 16 76 17 69 18 55 19 30 20 28 21 37 22 34 23 22 0 25 1 12 2 11 3 4 4 1 5 1 6 1 7 6 8 10 9 19 10 49 11 49 12 83 13 75 14 72 15 82 16 92 17 62 18 48 19 41 20 38 21 20 22 15 23 6 0 5 1 1 2 3 3 1 4 3 5 3 6 31 7 77 8 188 9 94 10 31 11 30 12 52 13 54 14 47 15 45 16 74 17 178 18 155 19 95 20 74 21 38 22 24 23 18 0 12 1 3 2 3 5 6 6 27 7 99 8 217 9 130 10 54 11 35 12 57 13 52 14 63 15 47 16 76 17 136 18 95 19 51 20 32 21 20 22 29 23 19 0 7 1 6 2 1 5 5 6 16 7 54 8 128 9 81 10 39 11 35 12 55 13 49 14 44 15 49 16 68 17 139 18 137 19 83 20 56 21 57 22 33 23 20 0 7 1 2 2 2 3 3 4 4 5 3 6 28 7 72 8 202 9 139 10 38 11 37 12 52 13 83 14 42 15 60 16 78 17 162 18 144 19 99 20 64 21 40 22 30 23 15 0 14 1 5 2 1 3 1 5 8 6 17 7 70 8 158 9 117 10 44 11 53 12 61 13 77 14 64 15 68 16 90 17 159 18 139 19 92 20 68 21 52 22 36 23 27 0 28 1 20 2 12 3 8 4 5 5 1 6 3 7 10 8 23 9 33 10 59 11 72 12 89 13 101 14 118 15 129 16 128 17 83 18 84 19 74 20 41 21 57 22 26 23 44 0 39 1 23 2 16 3 15 4 1 5 2 6 1 7 3 8 18 9 32 10 79 11 93 12 104 13 118 14 91 15 113 16 99 17 105 18 67 19 61 20 57 21 28 22 21 23 18 0 17 1 16 2 8 3 2 4 3 5 1 6 5 7 13 8 33 9 47 10 57 11 64 12 80 13 93 14 86 15 93 16 82 17 71 18 92 19 60 20 33 21 27 22 13 23 4 12 3 13 22 14 28 15 35 16 61 17 125 18 133 19 99 20 83 21 41 22 33 23 20 0 3 1 7 2 3 4 2 5 7 6 32 7 90 8 197 9 109 10 47 11 52 12 70 13 78 14 75 15 82 16 104 17 197 18 161 19 112 20 76 21 59 22 59 23 28 0 13 1 5 2 2 3 1 4 1 5 6 6 35 7 101 8 249 9 143 10 57 11 68 12 84 13 98 14 81 15 70 16 91 17 215 18 185 19 152 20 126 21 57 22 56 23 31 0 21 1 6 2 2 3 1 4 1 5 5 6 27 7 68 8 217 9 166 10 63 11 59 12 78 13 73 14 62 15 65 16 97 17 161 18 120 19 96 20 53 21 41 22 34 23 27 0 13 1 12 2 11 3 7 4 3 6 2 7 8 8 27 9 40 10 53 11 63 12 70 13 84 14 75 15 103 16 83 17 67 18 54 19 59 20 45 21 39 22 30 23 33 0 22 1 13 2 18 3 5 5 3 6 1 7 2 8 19 9 28 10 58 11 99 12 116 13 87 14 110 15 77 16 65 17 55 18 49 19 50 20 35 21 25 22 28 23 21 0 7 1 1 3 1 4 1 5 5 6 15 7 84 8 177 9 102 10 40 11 46 12 63 13 60 14 45 15 57 16 70 17 184 18 153 19 106 20 81 21 59 22 35 23 24 0 9 1 5 2 2 4 1 5 9 6 36 7 108 8 238 9 144 10 55 11 61 12 106 13 93 14 68 15 84 16 116 17 222 18 225 19 146 20 119 21 45 22 53 23 40 0 17 1 5 2 10 5 1 6 8 7 30 8 72 9 58 10 28 11 41 12 48 13 47 14 36 15 43 16 36 17 26 16 24 17 84 18 104 19 79 20 59 21 38 22 27 23 16 0 9 1 3 2 2 3 1 5 4 6 16 7 60 8 157 9 101 10 49 11 30 12 29 13 31 14 38 15 41 16 80 17 149 18 109 19 89 20 62 21 58 22 26 23 23 0 28 1 20 2 15 3 8 4 3 6 2 7 5 8 34 9 34 10 55 11 64 12 78 13 65 14 99 15 120 16 107 17 91 18 68 19 58 20 43 21 36 22 32 23 33 0 33 1 29 2 11 3 8 4 1 5 3 7 3 8 12 9 38 10 64 11 59 12 97 13 84 14 122 15 109 16 123 17 77 18 65 19 55 20 33 21 28 22 21 23 21 0 7 1 7 2 1 3 2 4 2 5 8 6 37 7 72 8 185 9 112 10 69 11 48 12 68 13 54 14 86 15 44 16 86 17 161 18 156 19 111 20 78 21 56 22 34 23 17 0 8 1 3 2 2 3 2 5 3 6 22 7 52 8 135 9 116 10 47 11 51 12 55 13 52 14 54 15 52 16 64 17 176 18 168 19 108 20 74 21 64 22 36 23 16 0 2 1 3 2 4 3 1 4 1 5 3 6 18 7 49 8 155 9

Create a scattermatrix.  This can be done in Python or with the built-in Databricks functionality.

In [48]:
%python
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

try:
  bikeDF
except NameError: # Looks for local table if bikeDF not defined
  bikeDF = spark.table("bike")

# use matplotlib subplots
fig, ax= plt.subplots()
# create a pandas dataframe from spark dataframe
pandasDF = bikeDF.select("*").toPandas()
# plot the pandas dataframe using scatter matrix
sm=pd.plotting.scatter_matrix(pandasDF)
[s.xaxis.label.set_rotation(90) for s in sm.reshape(-1)]
[s.yaxis.label.set_rotation(0) for s in sm.reshape(-1)]
[s.get_yaxis().set_label_coords(-0.99,0.15) for s in sm.reshape(-1)]
display(fig.figure)

### Step 3: Correlations

Calculate the correlations of the different variables.  Start by using `VectorAssembler` to put all the variables into a single column `features`.

In [50]:
# TODO
from pyspark.ml.feature import VectorAssembler
assembler = VectorAssembler(inputCols=bikeDF.columns,outputCol="features")
bikeFeaturizedDF=assembler.transform(bikeDF)

Calculate the correlation on the transformed `bikeDF`

In [52]:
# TODO
from pyspark.ml.stat import Correlation

pearsonCorr = Correlation.corr(bikeFeaturizedDF,"features").collect()[0][0]
pandasDF = pd.DataFrame(pearsonCorr.toArray())

pandasDF.index, pandasDF.columns = bikeDF.columns, bikeDF.columns

import matplotlib.pyplot as plt
import seaborn as sns

fig, ax = plt.subplots()
sns.heatmap(pandasDF)
display(fig.figure)

## Review
**Question:** Why is exploratory analysis important?  
**Answer:** The main goal of exploratory analysis is to build intuition from the dataset that will determine future modeling decisions.  This is often an overlooked step.  Having a limited understanding of the data leads to downstream problems like unexpected outputs and errors as well as under-performing models.

**Question:** What summary statistics are helpful?  
**Answer:** The most helpful statistics quantify the center and spread of the data (usually through the mean and standard deviation).  Counts (which reflect missing values) as well as the minimum and maximum are also informative.  A median can be used instead of a mean as well since a median is less influenced by outliers. 

**Question:** What are the most informative plots and how can I do that on Databricks?  
**Answer:** For a single variable, histograms and KDE plots are quite informative.  Box plots or violin plots are helpful as well.  Each shows the distribution of data and outliers in different ways.  When plotting multiple variables, scatterplots and bar charts can be helpful.  Scattermatrixes represent multiple features well.  This can be accomplished on Databricks either by using the built-in plotting functionality or by using the `display()` function to display plots generated with Python libraries like Matplotlib and Seaborn.

**Question:** What is correlation?  
**Answer:** Correlation looks at the linear relationship between two variables.  A positive correlation means that one variable increases linearly relative to another.  A negative correlation means that one variable increases as the other decreases.  No correlation means that the two variables are generally independent.

## Next Steps

Start the next lesson, [ML Workflows]($./04-ML-Workflows ).

## Additional Topics & Resources

**Q:** Where can I find out more information on general mathematical functions in Spark?  
**A:** Check out the Databricks blog post <a href="https://databricks.com/blog/2015/06/02/statistical-and-mathematical-functions-with-dataframes-in-spark.html" target="_blank">Statistical and Mathematical Functions with DataFrames in Spark</a>

**Q:** Where can I find out more information on the machine learning Spark library and statistics?  
**A:** Check out <a href="https://spark.apache.org/docs/latest/ml-statistics.html" target="_blank">the Spark documentation</a>

**Q:** Can I do kernel density estimation in Spark?  
**A:** Yes.  See the <a href="https://spark.apache.org/docs/latest/mllib-statistics.html#kernel-density-estimation" target="_blank">Spark documentation for details</a>

-sandbox
&copy; 2019 Databricks, Inc. All rights reserved.<br/>
Apache, Apache Spark, Spark and the Spark logo are trademarks of the <a href="http://www.apache.org/">Apache Software Foundation</a>.<br/>
<br/>
<a href="https://databricks.com/privacy-policy">Privacy Policy</a> | <a href="https://databricks.com/terms-of-use">Terms of Use</a> | <a href="http://help.databricks.com/">Support</a>